### PR TITLE
Allow using `Range` as an `Iterator` in const contexts. 

### PR DESCRIPTION
**Pull request by onestacked**
~~based on #102225 by @fee1-dead~~

### CODE REVIEW

**Comment by chriss0612** [comment]
#104113 should fix the build failure

**Comment by the8472** [comment]
Looks like #106541 is preferred over  #102225 now, so this will have to be rebased on that at some point.

**Comment by chriss0612** [comment]
I will probably only rebase once on of the implementations seems close to be merged.

**Comment by chriss0612** [comment]
rebased on top of https://github.com/rust-lang/rust/pull/106541

**Comment by chriss0612** [comment]
I think that this should be fine to merge now, but it might be better to wait for the bootstrap bump since it uses a bit of `#[cfg(bootstrap)]`.

**Comment by the8472** [comment]
Modulo the `#[cfg((bootstrap)]` this is small enough from a libs maintenance perspective. My understanding is that the `Destruct` stuff will go away at some point.

Handing off to @fee1-dead for the "do we want this in const eval" part (which probably is a rubber-stamp).

**Review by the8472 — APPROVED** [review]

File: tests/ui/consts/const_range_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    const {
+        let mut arr = [0; 3];
+        for i in 0..arr.len() {
+            arr[i] = i;
+        }
+        assert!(arr[0] == 0);
+        assert!(arr[1] == 1);
+        assert!(arr[2] == 2);
+    }
+}
```
Comment (fee1-dead) [inline tests/ui/consts/const_range_iterator.rs:19]: Please move this to libcore's unit tests.

**Review by fee1-dead — CHANGES_REQUESTED** [review]

File: tests/ui/consts/const_range_iterator.rs (hunk 1 above)
Comment (chriss0612) [reply tests/ui/consts/const_range_iterator.rs]: moved to library/core/tests/iter/consts.rs and added a test for a simple manually implemented `Iterator`.

**Comment by fee1-dead** [comment]
LGTM after fixing CI

**Comment by frewsxcv** [comment]
FYI this change broke the objc2 crate https://github.com/madsmtm/objc2/issues/432

**Comment by the8472** [comment]
You should open an issue for that, comments on merged PRs won't get much attention.

**Pull request merged**